### PR TITLE
introduce from parameter for mailserver requests

### PR DESCRIPTION
**Pull request by yenda**
- starts at current time so that nothing is fetched from mailserver when
recovering account
- still fetches history of a public chat when joining

Steps to test:
- create new account, open public chat, history should be fetched
- restore an account, nothing should be fetched
- be @chadyj  and have a slow client, update to this version and check if it's faster

Part of #4387

status: ready

### CODE REVIEW

**Review by dmitryn — APPROVED** [review]

**Comment by rasom** [comment]
apk uploaded to https://i.diawi.com/G7ebYn for branch PR-4230

<blockquote><div><strong><a href="https://i.diawi.com/G7ebYn">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by rasom** [comment]
ipa uploaded to https://i.diawi.com/E3gohe for branch PR-4230

<blockquote><div><strong><a href="https://i.diawi.com/E3gohe">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by chadyj** [comment]
Thanks Eric!
I tested this out and lost some older messages that were sent since the last time I used the app.  
I'll leave it installed over the weekend and see how it feels.

**Review by janherich — APPROVED** [review]

**Comment by yenda** [comment]
@chadyj it is expected that you don't receive messages that were sent before you update

**Comment by rasom** [comment]
apk uploaded to https://i.diawi.com/TrkBFh for branch PR-4230

<blockquote><div><strong><a href="https://i.diawi.com/TrkBFh">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by rasom** [comment]
ipa uploaded to https://i.diawi.com/8wSxff for branch PR-4230

<blockquote><div><strong><a href="https://i.diawi.com/8wSxff">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

File: src/status_im/transport/inbox.cljs [hunk 1]
```diff
@@ -219,18 +219,25 @@
...
-                          :web3       web3}
-      :db (assoc db :inbox/last-request (quot now 1000))})))
+         from       (or from (:last-request account))
+         sym-key-id (:inbox/sym-key-id db)
+         now-in-s   (quot now 1000)
+         fx         {:data-store/save-account (assoc account :last-request now-in-s)
+                     :db                      (assoc-in db [:account/account :last-request] now-in-s)}]
+     (if from
+       ;; NOTE (yenda) we ensure a delay of 10 seconds between mailserver requests
+       (when (< 10 (- now-in-s from))
```
Comment (rasom) [inline src/status_im/transport/inbox.cljs]: @yenda please extract magic value and give it name

File: src/status_im/transport/inbox.cljs [hunk 2]
```diff
@@ -219,18 +219,25 @@
...
+         (merge fx
+                {::request-messages       {:wnode      wnode
+                                           :topics     (if discover?
+                                                         (conj topics
+                                                               (transport.utils/get-topic constants/contact-discovery))
```
Comment (rasom) [inline src/status_im/transport/inbox.cljs]: 80 chars limit

File: src/status_im/transport/inbox.cljs [hunk 3]
```diff
@@ -219,18 +219,25 @@
...
+         from       (or from (:last-request account))
+         sym-key-id (:inbox/sym-key-id db)
+         now-in-s   (quot now 1000)
+         fx         {:data-store/save-account (assoc account :last-request now-in-s)
+                     :db                      (assoc-in db [:account/account :last-request] now-in-s)}]
+     (if from
+       ;; NOTE (yenda) we ensure a delay of 10 seconds between mailserver requests
+       (when (< 10 (- now-in-s from))
+         (merge fx
+                {::request-messages       {:wnode      wnode
```
Comment (rasom) [inline src/status_im/transport/inbox.cljs]: also map alignment is a bit weird here

**Review by rasom — APPROVED** [review]

**Comment by rasom** [comment]
apk uploaded to https://i.diawi.com/HKvcTr for branch PR-4230

<blockquote><div><strong><a href="https://i.diawi.com/HKvcTr">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by rasom** [comment]
ipa uploaded to https://i.diawi.com/GsuPAH for branch PR-4230

<blockquote><div><strong><a href="https://i.diawi.com/GsuPAH">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

File: src/status_im/transport/inbox.cljs [hunk 4]
```diff
@@ -13,6 +13,8 @@
             [day8.re-frame.async-flow-fx]
             [status-im.constants :as constants]))
 
+(def delay-in-s-between-mailserver-requests 10)
```
Comment (cammellos) [inline src/status_im/transport/inbox.cljs:16]: What is the reason of using 10? Do we get disconnected from the mailserver if we make more than 1 request every x seconds or it's just arbitrarily chosen to prevent sending too many requests? (Essentially if it's something that the mailserver requires I would leave a comment to document it)

File: src/status_im/transport/inbox.cljs [hunk 5]
```diff
@@ -219,18 +221,22 @@
...
-                          :sym-key-id sym-key-id
-                          :web3       web3}
-      :db (assoc db :inbox/last-request (quot now 1000))})))
+         topics     (cond-> (or topics
+                                (map #(:topic %) (vals (:transport/chats db))))
+                      discover? (conj (transport.utils/get-topic constants/contact-discovery)))
+         from       (or from (:last-request account))
+         sym-key-id (:inbox/sym-key-id db)
+         now-in-s   (quot now 1000)
+         fx         {:data-store/save-account (assoc account :last-request now-in-s)
```
Comment (cammellos) [inline src/status_im/transport/inbox.cljs:232]: It seems here we store  `:last-request` even if we:

1) Don't make the request because it comes in less than 10 seconds before the last
2) The request to the mailserver fails

In both cases we might lose some messages.

Seems like 1 it's easy to change, unless there are some other reason why we do that?

2 Is a bit harder, we should probably either store in only on success (which brings some complications), or if we want to store it optimistically like this then probably worth retrying the request a few times, give up and move on. 
Maybe not necessary in this PR but worth leaving a comment as something we need to address.

File: src/status_im/transport/inbox.cljs [hunk 6]
```diff
@@ -219,18 +221,22 @@
...
+                     :db                      (assoc-in db [:account/account :last-request] now-in-s)}]
+     (if from
+       (when (< delay-in-s-between-mailserver-requests (- now-in-s from))
```
Comment (cammellos) [inline src/status_im/transport/inbox.cljs:235]: As I understand by looking at the code, if we poll the mailserver more than once every 10s we drop the request ( also we reset the counter, which probably should not do). 

Not sure if this is a limitation of the mailserver ( 1 req per 10s), but this might occur in behaviours like:

1) I receive a contact request, poll the mailserver
2) I join a public chat after 5s, no history is fetched

or 

1) I go online, fetch from the mailserver
2) I receive a contact request, no history is fetched

and so on.

It would work if all the requests to the mailserver were identical, but because we can specify a topic, that's not the case.

Why are we limiting the requests, is it something on the mailserver (we get disconnected if more than 1 request per 10 seconds for example), or we are just trying to prevent sending to many requests? (What might be the cause of this, flip-flop of network connection? or too many contact requests/joining public chats?)

File: src/status_im/transport/message/v1/contact.cljs [hunk 7]
```diff
@@ -76,8 +76,8 @@
     (let [message-id (transport.utils/message-id this)
           public-keys (remove nil? (map :public-key (vals (:contacts/contacts db))))]
       (handlers-macro/merge-fx cofx
-                               (protocol/multi-send-with-pubkey {:public-keys public-keys
-                                                                 :payload     this}))))
+                               (protocol/multi-send-by-pubkey {:public-keys public-keys
```
Comment (cammellos) [inline src/status_im/transport/message/v1/contact.cljs:79]: Looks like these changes have been merged already or not related to this PR?

File: src/status_im/transport/inbox.cljs [hunk 8]
```diff
@@ -219,18 +221,22 @@
...
-                          ;;TODO (yenda) fix from, right now mailserver is dropping us
-                          ;;when we send a requestMessage with a from field
-                          ;;:from       from
-                          :sym-key-id sym-key-id
-                          :web3       web3}
-      :db (assoc db :inbox/last-request (quot now 1000))})))
+         topics     (cond-> (or topics
+                                (map #(:topic %) (vals (:transport/chats db))))
+                      discover? (conj (transport.utils/get-topic constants/contact-discovery)))
+         from       (or from (:last-request account))
```
Comment (cammellos) [inline src/status_im/transport/inbox.cljs:229]: Seems like we use a single cursor for different topics:

`last-request` is 20
I ask for topic `B` with from = 20, now `last-request` is 30
Next time I query the mailserver, I ask for all the topics with from = 30.

Seems to me we should ask from 20 the second time as well, otherwise we might lose some messages.

This makes the system quite difficult to reason about.

Two ways to handle this I can see, either we have a cursor for each topic + discovery (something like `transport/chats chat-id last-request`), or we always fetch all the topics. What do you think?

**Review by cammellos — CHANGES_REQUESTED** [review]
Looks like a good start, but needs a bit more working, I would either make it as simple as possible (fetch always all topics ), or go for a more robust implementation (keep a cursor for each topic). Or we could think of other ways, what do you reckon?

File: src/status_im/transport/inbox.cljs (hunk 6 above)
Comment (yenda) [reply src/status_im/transport/inbox.cljs:235]: We don't update the counter when we drop the request because it's done less than 10 sec since the last call to the mailserver it's a when and nothing happens when it's false.

the condition should be more sophisticated though to fetch every time and not update when there is a topic specified.

This delay avoids spamming the mailserver on flaky connections.

I have thought about other ways to do this over the weekend but there is no easy solution, having one clock per topic wouldn't provide much benefits and would increase complexity/might not even be doable in the future if we send a bloomfilter as a request. 

The case of failing requests is the most problematic as there is no real success for requests, you can't tell if you have received all the messages you requested before losing connectivity for instance.

I think the performance benefits of this PR are on par with the deduplication PR on go side and only saves more bandwitdth. We can probably delay it to after beta so that we can think more of improvements to request message and consequences of using from parameter

File: src/status_im/transport/message/v1/contact.cljs (hunk 7 above)
Comment (yenda) [reply src/status_im/transport/message/v1/contact.cljs:79]: yes I made an odd commit

File: src/status_im/transport/inbox.cljs (hunk 6 above)
Comment (cammellos) [reply src/status_im/transport/inbox.cljs:235]: Ah you are correct, I got my brackets mixed up! So yes, the only problem would be if mailserver if the connection is down/mailserver is down. But yes, as it is we might lose some messages, so not sure whether we want to include it. What about keeping a single counter and always polling all the topics? That should take care of some edge cases at least, but would also means we don't pull the whole history for a public chat (I remember the test team mentioned we should not fetch history when joining a public chat).

File: src/status_im/transport/inbox.cljs (hunk 8 above)
Comment (pedropombeiro) [reply src/status_im/transport/inbox.cljs:229]: > Two ways to handle this I can see, either we have a cursor for each topic + discovery (something like transport/chats chat-id last-request), or we always fetch all the topics. What do you think?

@yenda what do you think regarding the above question?

Comment (cammellos) [reply src/status_im/transport/inbox.cljs:229]: there's also the option as mentioned to don't update the cursor when you make one-off requests (specify a topic), only when we ask for all the topics we update it.

**Comment by jeluard** [comment]
@yenda What is the state of this PR?

**Comment by pedropombeiro** [comment]
@yenda which GHI is this PR addressing?

**Comment by yenda** [comment]
@PombeirP it's a blocker for https://github.com/status-im/status-react/issues/4387. I will revisit it after signal handling is merged. Currently fixing a bug on it
